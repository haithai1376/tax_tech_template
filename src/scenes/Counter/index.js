import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export default class Counter extends React.Component {
  static propTypes = {
    status: PropTypes.string.isRequired,
    count: PropTypes.number.isRequired,
    updateCount: PropTypes.func.isRequired,
  }

  state = {
    animate: false,
  }

  componentWillReceiveProps(nextProps) {
    const { count: currentCount } = this.props;
    const { count: newCount } = nextProps;

    if (currentCount != newCount) {
      this.setState({ animate: true });
      setTimeout(() => {
        this.setState({ animate: false });
      }, 250);
    }
  }

  increase = () => {
    const {
      count,
      updateCount,
    } = this.props;

    updateCount(count + 1);
  }

  decrease = () => {
    const {
      count,
      updateCount,
    } = this.props;

    updateCount(count - 1);
  }

  render() {
    const {
      animate
    } = this.state;
    const {
      status,
      count,
    } = this.props;

    return (
      <StyledCounter>
        <Display fontSize="1rem">{status}</Display>
        <Display fontSize="6rem" animate={animate}>{count}</Display>
        <ButtonGroup>
          <Button bgColor="#d0d0d0" onClick={this.increase}>+</Button>
          <Button bgColor="#a0a0a0" onClick={this.decrease}>-</Button>
        </ButtonGroup>
      </StyledCounter>
    );
  }
}

export const StyledCounter = styled.div`
  margin: auto;
  display: flex;
  flex-direction: column;
  align-self: center;
`;

export const Display = styled.div`
  font-size: ${props => props.fontSize};
  color: ${props => props.animate ? '#61dafb' : '#282c34'};
  background-color: #ffffff;
  text-align: center;
  margin-bottom: 1px;
  padding: 1rem;
  transition: color .25s;
`;

export const ButtonGroup = styled.div`
  display: flex;
`;

export const Button = styled.div`
  font-size: 3rem;
  color: #ffffff;
  background-color: #282c34;
  width: 100px;
  text-align: center;
  transition: background-color .25s;

  &:hover {
    color: #ffffff;
    background-color: #61dafb;
    cursor: pointer;
  }
`;
