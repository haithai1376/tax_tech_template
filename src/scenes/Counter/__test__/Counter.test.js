import React from 'react';
import styled from 'styled-components';
import { shallow, mount, render } from 'enzyme';
import Counter, { ButtonGroup, Button } from '..';

describe('Counter Component', () => {
  const props = {
    status: 'Testing',
    count: 0,
    updateCount: jest.fn(),
  };

  test('render', () => {
    const wrapper = shallow(<Counter {...props}/>);
    expect(wrapper).toMatchSnapshot();
  });

  test('receive same props', () => {
    jest.useFakeTimers();
    const wrapper = mount(<Counter {...props}/>);
    wrapper.setProps({ count: 0 });
    jest.runAllTimers();
    expect(wrapper.prop('count')).toBe(0);
  });

  test('receive new props', () => {
    jest.useFakeTimers();
    const wrapper = mount(<Counter {...props}/>);
    wrapper.setProps({ count: 1 });
    jest.runAllTimers();
    expect(wrapper.prop('count')).toBe(1);
  });

  test('click increment button', () => {
    const wrapper = mount(<Counter {...props}/>);
    const spy = jest.spyOn(wrapper.instance(), 'increase');
    wrapper.instance().forceUpdate();
    wrapper.find(ButtonGroup).find(Button).at(0).simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  test('click decrement button', () => {
    const wrapper = mount(<Counter {...props}/>);
    const spy = jest.spyOn(wrapper.instance(), 'decrease');
    wrapper.instance().forceUpdate();
    wrapper.find(ButtonGroup).find(Button).at(1).simulate('click');
    expect(spy).toHaveBeenCalled();
  });
});
