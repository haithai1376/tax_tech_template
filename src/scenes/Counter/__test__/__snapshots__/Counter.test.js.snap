// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Counter
    count={0}
    status="Testing"
    updateCount={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styled.div
          fontSize="1rem"
        >
          Testing
        </styled.div>,
        <styled.div
          animate={false}
          fontSize="6rem"
        >
          0
        </styled.div>,
        <styled.div>
          <styled.div
            bgColor="#d0d0d0"
            onClick={[Function]}
          >
            +
          </styled.div>
          <styled.div
            bgColor="#a0a0a0"
            onClick={[Function]}
          >
            -
          </styled.div>
        </styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Testing",
          "fontSize": "1rem",
        },
        "ref": null,
        "rendered": "Testing",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animate": false,
          "children": 0,
          "fontSize": "6rem",
        },
        "ref": null,
        "rendered": 0,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <styled.div
              bgColor="#d0d0d0"
              onClick={[Function]}
            >
              +
            </styled.div>,
            <styled.div
              bgColor="#a0a0a0"
              onClick={[Function]}
            >
              -
            </styled.div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bgColor": "#d0d0d0",
              "children": "+",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "+",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bgColor": "#a0a0a0",
              "children": "-",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "-",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.div
            fontSize="1rem"
          >
            Testing
          </styled.div>,
          <styled.div
            animate={false}
            fontSize="6rem"
          >
            0
          </styled.div>,
          <styled.div>
            <styled.div
              bgColor="#d0d0d0"
              onClick={[Function]}
            >
              +
            </styled.div>
            <styled.div
              bgColor="#a0a0a0"
              onClick={[Function]}
            >
              -
            </styled.div>
          </styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Testing",
            "fontSize": "1rem",
          },
          "ref": null,
          "rendered": "Testing",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "animate": false,
            "children": 0,
            "fontSize": "6rem",
          },
          "ref": null,
          "rendered": 0,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <styled.div
                bgColor="#d0d0d0"
                onClick={[Function]}
              >
                +
              </styled.div>,
              <styled.div
                bgColor="#a0a0a0"
                onClick={[Function]}
              >
                -
              </styled.div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bgColor": "#d0d0d0",
                "children": "+",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "+",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bgColor": "#a0a0a0",
                "children": "-",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "-",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
